File: implementations/ios-sdk/OptimizationAppUITests/Tests/PreviewPanelTests.swift
import XCTest

final class PreviewPanelTests: XCTestCase {
    let app = XCUIApplication()

    override func setUp() {
        continueAfterFailure = false
        app.launch()
        clearProfileState(app: app)
        waitForElement(app.buttons["identify-button"])
    }

    // MARK: - Helpers

    private func openPreviewPanel() {
        let fab = app.buttons["preview-panel-fab"]
        waitForElement(fab)
        fab.tap()

        let title = app.staticTexts["Preview Panel"]
        XCTAssertTrue(title.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Preview Panel sheet did not appear")
    }

    /// Scrolls the preview panel list until the target element is visible.
    /// SwiftUI List renders as collectionView; plain ScrollView as scrollView.
    /// Prefer an identifier match; fall back to the first match of either type.
    private func scrollToPreviewElement(_ testId: String, maxSwipes: Int = 10) {
        let listById = app.collectionViews["preview-panel-list"]
        let scrollById = app.scrollViews["preview-panel-list"]
        let scrollContainer: XCUIElement
        if listById.exists {
            scrollContainer = listById
        } else if scrollById.exists {
            scrollContainer = scrollById
        } else if app.collectionViews.firstMatch.exists {
            scrollContainer = app.collectionViews.firstMatch
        } else {
            scrollContainer = app.scrollViews.firstMatch
        }
        for _ in 0..<maxSwipes {
            let target = findElement(testId, app: app)
            if target.exists && target.isHittable { return }
            scrollContainer.swipeUp()
        }
    }

    // MARK: - FAB Visibility

    func testFABIsVisible() {
        let fab = app.buttons["preview-panel-fab"]
        XCTAssertTrue(fab.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Preview panel FAB should be visible on main screen")
    }

    // MARK: - Profile Data Loading

    func testShowsProfileData() {
        openPreviewPanel()

        let noData = app.staticTexts["no-profile-data"]
        _ = noData.waitForExistence(timeout: 2.0)
        XCTAssertFalse(noData.exists,
                       "Should not show 'No profile data' after initialization")
    }

    func testShowsAllExpectedProfileKeys() {
        openPreviewPanel()

        let expectedKeys = ["audiences", "id", "location", "random", "session", "stableId", "traits"]
        for key in expectedKeys {
            scrollToPreviewElement("profile-item-\(key)")
            let item = findElement("profile-item-\(key)", app: app)
            XCTAssertTrue(item.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                          "Expected profile key '\(key)' not found in preview panel")
        }
    }

    func testShowsLocationData() {
        openPreviewPanel()

        scrollToPreviewElement("profile-item-location")
        let locationItem = findElement("profile-item-location", app: app)
        XCTAssertTrue(locationItem.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Profile location item not found")

        let label = extractText(locationItem)
        XCTAssertTrue(label.contains("Berlin") || label.contains("DE"),
                      "Expected location to contain Berlin or DE, got: \(label)")
    }

    // MARK: - Debug Section

    func testShowsConsentAccepted() {
        openPreviewPanel()

        scrollToPreviewElement("debug-consent")
        let consent = findElement("debug-consent", app: app)
        XCTAssertTrue(consent.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Debug consent element not found")
        let text = extractText(consent)
        XCTAssertTrue(text.contains("Accepted"),
                      "Expected consent to contain 'Accepted', got: \(text)")
    }

    func testShowsCanPersonalize() {
        openPreviewPanel()

        scrollToPreviewElement("debug-can-personalize")
        let canPersonalize = findElement("debug-can-personalize", app: app)
        XCTAssertTrue(canPersonalize.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Debug canPersonalize element not found")
        let text = extractText(canPersonalize)
        XCTAssertTrue(text.contains("Yes"),
                      "Expected canPersonalize to contain 'Yes', got: \(text)")
    }

    // MARK: - Refresh

    func testRefreshButtonWorks() {
        openPreviewPanel()

        scrollToPreviewElement("preview-refresh-button")
        let refreshButton = app.buttons["preview-refresh-button"]
        XCTAssertTrue(refreshButton.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Refresh button should exist")
        refreshButton.tap()

        let noData = app.staticTexts["no-profile-data"]
        _ = noData.waitForExistence(timeout: 2.0)
        XCTAssertFalse(noData.exists,
                       "Profile data should persist after refresh")
    }

    // MARK: - Profile After Identify

    func testProfileUpdatesAfterIdentify() {
        app.buttons["identify-button"].tap()
        waitForElement(app.buttons["reset-button"])

        openPreviewPanel()

        let noData = app.staticTexts["no-profile-data"]
        _ = noData.waitForExistence(timeout: 2.0)
        XCTAssertFalse(noData.exists,
                       "Should show profile data after identify")

        let idItem = findElement("profile-item-id", app: app)
        XCTAssertTrue(idItem.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Profile id should be present after identify")
    }
}

File: implementations/android-sdk/uitests/src/main/kotlin/com/contentful/optimization/uitests/tests/PreviewPanelTests.kt
package com.contentful.optimization.uitests.tests

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.By
import androidx.test.uiautomator.UiDevice
import com.contentful.optimization.uitests.support.AppLauncher
import com.contentful.optimization.uitests.support.PerTestRule
import com.contentful.optimization.uitests.support.TestHelpers
import com.contentful.optimization.uitests.support.clearProfileState
import org.junit.Assert
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class PreviewPanelTests {
    @get:Rule
    val rule: TestRule = PerTestRule.create()

    private lateinit var device: UiDevice

    @Before
    fun setUp() {
        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
        AppLauncher.launchApp(device)
        clearProfileState(device)
        TestHelpers.waitForElement(device, By.res("identify-button"))
    }

    private fun openPreviewPanel() {
        TestHelpers.waitAndTap(device, By.desc("preview-panel-fab"))
        TestHelpers.waitForElement(device, By.text("Preview Panel"))
        Thread.sleep(2000)
    }

    private fun scrollToPreviewElement(testId: String, maxSwipes: Int = 20) {
        for (i in 0 until maxSwipes) {
            if (TestHelpers.findElement(device, testId) != null) return
            val panelBounds = device.findObject(By.desc("preview-panel-list"))?.visibleBounds
            val centerX = panelBounds?.centerX() ?: (device.displayWidth / 2)
            val startY = panelBounds?.let { it.top + (it.height() * 3 / 4) } ?: (device.displayHeight * 3 / 4)
            val endY = panelBounds?.let { it.top + (it.height() / 4) } ?: (device.displayHeight / 4)
            device.swipe(centerX, startY, centerX, endY, 10)
            Thread.sleep(300)
        }
    }

    // FAB Visibility

    @Test
    fun testFABIsVisible() {
        val fab = device.findObject(By.desc("preview-panel-fab"))
        Assert.assertNotNull("Preview panel FAB should be visible on main screen", fab)
    }

    // Profile Data Loading

    @Test
    fun testShowsProfileData() {
        openPreviewPanel()
        Thread.sleep(2000)
        val noData = device.findObject(By.desc("no-profile-data"))
        Assert.assertNull("Should not show 'No profile data' after initialization", noData)
    }

    @Test
    fun testShowsAllExpectedProfileKeys() {
        openPreviewPanel()

        val expectedKeys = listOf("audiences", "id", "location", "random", "session", "stableId", "traits")
        for (key in expectedKeys) {
            scrollToPreviewElement("profile-item-$key")
            val item = TestHelpers.findElement(device, "profile-item-$key")
            Assert.assertNotNull("Expected profile key '$key' not found in preview panel", item)
        }
    }

    @Test
    fun testShowsLocationData() {
        openPreviewPanel()
        scrollToPreviewElement("profile-item-location")

        val locationItem = TestHelpers.findElement(device, "profile-item-location")
        Assert.assertNotNull("Profile location item not found", locationItem)

        val text = TestHelpers.extractText(locationItem!!)
        Assert.assertTrue(
            "Expected location to contain Berlin or DE, got: $text",
            text.contains("Berlin") || text.contains("DE"),
        )
    }

    // Debug Section

    @Test
    fun testShowsConsentAccepted() {
        openPreviewPanel()
        scrollToPreviewElement("debug-consent")

        val consent = TestHelpers.findElement(device, "debug-consent")
        Assert.assertNotNull("Debug consent element not found", consent)

        val text = TestHelpers.extractText(consent!!)
        Assert.assertTrue(
            "Expected consent to contain 'Accepted', got: $text",
            text.contains("Accepted"),
        )
    }

    @Test
    fun testShowsCanOptimize() {
        openPreviewPanel()
        scrollToPreviewElement("debug-can-optimize")

        val canOptimize = TestHelpers.findElement(device, "debug-can-optimize")
        Assert.assertNotNull("Debug canOptimize element not found", canOptimize)

        val text = TestHelpers.extractText(canOptimize!!)
        Assert.assertTrue(
            "Expected canOptimize to contain 'Yes', got: $text",
            text.contains("Yes"),
        )
    }

    // Refresh

    @Test
    fun testRefreshButtonWorks() {
        openPreviewPanel()
        scrollToPreviewElement("preview-refresh-button")

        TestHelpers.waitAndTap(device, By.desc("preview-refresh-button"))

        Thread.sleep(2000)
        val noData = device.findObject(By.desc("no-profile-data"))
        Assert.assertNull("Profile data should persist after refresh", noData)
    }

    // Profile After Identify

    @Test
    fun testProfileUpdatesAfterIdentify() {
        TestHelpers.waitAndTap(device, By.res("identify-button"))
        TestHelpers.waitForElement(device, By.res("reset-button"), TestHelpers.EXTENDED_TIMEOUT)

        openPreviewPanel()

        Thread.sleep(2000)
        val noData = device.findObject(By.desc("no-profile-data"))
        Assert.assertNull("Should show profile data after identify", noData)

        scrollToPreviewElement("profile-item-id")
        val idItem = TestHelpers.findElement(device, "profile-item-id")
        Assert.assertNotNull("Profile id should be present after identify", idItem)
    }
}
